dashboard/app/controllers/aichat_controller.rb

```diff
@@ -44,4 +46,66 @@ def has_required_params?
     # Check storedMessages param separately.
     params[:storedMessages].is_a?(Array)
   end
+
+  private def log_chat_session(assistant_message)
+    if params[:sessionId].present?
+      session_id = params[:sessionId]
+      session = AichatSession.find_by(id: session_id)
+      if session && matches_existing_session?(session)
+        return update_session(session, assistant_message)
+      end
+    end
+
+    create_session(assistant_message)
+  end
+
+  private def matches_existing_session?(session)
+    context = params[:aichatContext]
+    if session.level_id != context[:currentLevelId] ||
+        session.script_id != context[:scriptId] ||
+        current_user.id != session.user_id
+      return false
+    end
+
+    _, project_id = storage_decrypt_channel_id(context[:channelId])
+    if session.project_id != project_id
+      return false
+    end
+
+    if params[:aichatModelCustomizations] != JSON.parse(session.model_customizations) ||
+        params[:storedMessages] != JSON.parse(session.messages)
+      return false
+    end
+
+    true
+  end
+
+  private def update_session(session, assistant_message)
+    session.messages = updated_message_list(assistant_message).to_json
+    session.save
```

Comment:
Question - I've sometimes seen models saved with `.save!` which seems like it throws an error if saving fails. I'm guessing we don't want to do that here since we don't want to disrupt the user if the message sent successfully? But in that case should we do something with the return value (true/false) of `session.save`?

Reply:
Yeah good q, I was thinking we probably need an error handling pass here (I had a few notes in the description, but didn't explicitly cover the case you're mentioning). I was wondering how we want to handle if logging isn't working for whatever reason, do we want to treat that as the whole service is down (ie, don't respond to the student)? I could see that, if we want to be sure that we have a record of every chat interaction that happens. I could also see considering logging as non-blocking, so curious for your opinions.

Reply:
Relatedly, thinking about profanity filtering...I do think we'd want to log any profane input from the student, even if we filter it before it gets sent to Sagemaker. So will have to think on how we manage keeping that in the chat history, but excluded from the conversation history that we send to Sagemaker.

Reply:
Interesting thought! I wonder when you begin working on profanity filtering if it would make sense to store messages as a hash with content and status. I see that you've already typed as `:json`.

Reply:
> if logging isn't working for whatever reason, do we want to treat that as the whole service is down (ie, don't respond to the student)? I could see that, if we want to be sure that we have a record of every chat interaction that happens.

I agree that it seems safer to treat as the whole service not working if there was a saving error.

Reply:
_> Interesting thought! I wonder when you begin working on profanity filtering if it would make sense to store messages as a hash with content and status. I see that you've already typed as :json._

Yeah we're already storing each message as a hash with keys of `role` and `content`, might need to include a `status` key or something to handle profanity filtering...will noodle on it (and may have some questions for you based on your initial work here 🙏 )

Reply:
Re: error handling, yeah I'm a bit torn because a failure of logging is not necessary a fatal error if the student can continue chatting with the chatbot, but it would result in missed messages. I'd probably defer to @dancodedotorg / @samantha-code and ask, is it more important to try for a seamless student experience, or that we ensure that every chat interaction is logged, even if it means interruptions to students?

Reply:
^ took that question to Slack for Sam's input